src/input-states.js

```diff
@@ -0,0 +1,236 @@
+    else if (vnid) return getPortState(node, '', portName);
+    else return undefined;
+}
+
+/**
+ * Returns an array of the port states, one item for each attached socket. When
+ * all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortStateArray(node, vnid, portName) {
+    var port = node.inPorts[portName];
+    if (!isMultiPort(port)) throw Error("This port is not addressable/multi: " + portName);
+    else return port.listAttached().map(function(socketId){
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    });
+}
+
+/**
+ * Returns the state of this port for the given socketId. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketId is not neccessarily the same as the array index
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the name of the addressable/multi port for this node
+ * @param socketId the socketId provided to a port data event
+ */
+function getPortStateBySocketId(node, vnid, portName, socketId) {
+    if (socketId == null)
+        return getPortState(node, vnid, portName);
+    else if (!isMultiPort(node.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    else if (vniStateExists(node, vnid, portName) &&
+            node.vnis[vnid].inputStates[portName][socketId])
+        return node.vnis[vnid].inputStates[portName][socketId];
+    // By default states that come in on vnid '' apply to all VNIs
+    else if (vnid) return getPortStateBySocketId(node, '', portName, socketId);
+    else return undefined;
+}
+
+/**
+ * Changes the state of each given port. If a port name is missing it is unmodified.
+ * This can be run in a test environment to reset the state of the ports.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portStates hash of portName to state
+ */
+function setAllPortStates(node, vnid, portStates) {
+    _.each(portStates, function(state, portName) {
+        setPortState(node, vnid, portName, state);
+    });
+    return node;
+}
+
+/**
+ * Changes the state of a port. When a new state is received on a port this
+ * should be called.
+ * setPortStateBySocketId should be used to set the state on a multi/addressable port.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ * @param state the new state (or array of states if the port is multi/addressable) to assign to this port
+ */
+function setPortState(node, vnid, portName, state) {
+    var port = node.inPorts[portName];
+    if (_.isUndefined(state)) {
```

Comment:
Let's discuss this.  I don't think the state entry should be deleted from the inputStates hash.  But we do need a way to destroy a state and have the wrapper's destroyState method called -- maybe not in this module though.

As I see it, this function would normally only be used for testing -- not for runtime use -- because at runtime setPortStateBySocketId would be called to set a single state when an event comes in.  In which case, the comment above should be updated to indicate that this function would normally only be used for testing.